<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A CAusal Reasoning tool for Network Identification (from gene expression data) using Integer VALue programming • CARNIVAL</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="A CAusal Reasoning tool for Network Identification (from gene expression data) using Integer VALue programming">
<meta property="og:description" content="An upgraded causal reasoning tool from Melas et al in R with updated assignments of TFs' weights from PROGENy scores.
             Optimization parameters can be freely adjusted and multiple solutions can be obtained and aggregated.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">CARNIVAL</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/CARNIVAL.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/saezlab/CARNIVAL/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="carnival">CARNIVAL<a class="anchor" aria-label="anchor" href="#carnival"></a>
</h1></div>
<!-- badges: start -->

<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>CARNIVAL (<strong>CA</strong>usal <strong>R</strong>easoning for
<strong>N</strong>etwork identification using <strong>I</strong>nteger
<strong>VAL</strong>ue programming) is a method for the identification
of upstream reguatory signalling pathways from downstream gene
expression (GEX).</p>
<p>This is a tool currently being developed by the <a href="http://saezlab.org/" class="external-link">saezlab</a> members and is an extension of
the previously implemented Causal Reasoning (<a href="http://pubs.rsc.org/en/content/articlehtml/2015/ib/c4ib00294f" class="external-link"><em>Melas
et al.</em></a>) method. More detailed information on the CARNIVAL
pipeline as well as benchmarking and applicational studies are available
on following paper.</p>
<blockquote>
<p>Liu A., Trairatphisan P., Gjerga E. et al. <a href="https://www.nature.com/articles/s41540-019-0118-z" class="external-link">From expression
footprints to causal pathways: contextualizing large signaling networks
with CARNIVAL</a> <em>npj Systems Biology and Applications</em> volume
5, Article number: 40 (2019) (equal contributions).</p>
</blockquote>
<p>The aim of the CARNIVAL pipeline is to identify a subset of
interactions from a prior knowledge network that represent potential
regulated pathways linking known or potential targets of perturbation
towards active transcription factors derived from GEX data. The pipeline
includes a number improved functionalities comparing to the original
version and consists of the following processes:</p>
<ul>
<li><ol style="list-style-type: decimal">
<li>Transcription factors’ (TFs) activities and pathway scores from gene
expressions can be inferred with our in-house tools DoRothEA &amp;
PROGENy, respectively.</li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>TFs’ activities and signed directed protein-protein interaction
networks with or without the provided target of perturbations and
pathway scores are then used to derive a series of linear constraints to
generate integer linear programming (ILP) problems.</li>
</ol></li>
<li><ol start="3" style="list-style-type: decimal">
<li>An ILP solver (IBM ILOG CPLEX) is subsequently applied to identify
the sub-network topology with minimised fitting error and model
size.</li>
</ol></li>
</ul>
<p>Applications of CARNIVAL include the identification of drug’s modes
of action and of deregulated processes in diseases (even if the
molecular targets remain unknown) by deciphering the alterations of main
signalling pathways as well as alternative pathways and off-target
effects.</p>
<p><img src="reference/figures/graphical_abstract.png" align="center" width="800"></p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting Started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>A tutorial for preparing CARNIVAL input files starting from
differentially gene expression (DEG) and for running the CARNIVAL
pipeline are provided as vignettes in R-Markdown, R-script and HTML
formats. The wrapper script “runCARNIVAL” was introduced to take input
arguments, pre-process input descriptions, run optimisation and export
results as network files and figures. Three built-in CARNIVAL examples
are also supplied as case studies for users.</p>
<div class="section level3">
<h3 id="prerequisites">Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h3>
<p>CARNIVAL requires the interactive version of IBM Cplex, Gurobi or
CBC-COIN solver as the network optimiser. The IBM ILOG Cplex is freely
available through Academic Initiative <a href="https://www.ibm.com/products/ilog-cplex-optimization-studio" class="external-link">here</a>.
Gurobi license is also free for academics, request a license <a href="https://www.gurobi.com/downloads/end-user-license-agreement-academic/" class="external-link">here</a>.
The <a href="https://projects.coin-or.org/Cbc" class="external-link">CBC</a> solver is open
source and freely available for any user, but has a significantly lower
performance than CPLEX or Gurobi. Obtain CBC executable directly usable
for CARNIVAL <a href="https://ampl.com/products/solvers/open-source/#cbc" class="external-link">here</a></p>
<p>Alternatively for small networks, users can rely on the freely
available <a href="https://cran.r-project.org/web/packages/lpSolve/index.html" class="external-link">lpSolve
R-package</a>, which is automatically installed with the package.</p>
</div>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>To install the stable version from <a href="http://bioconductor.org/" class="external-link">Bioconductor</a>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># install from bioconductor</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"CARNIVAL"</span><span class="op">)</span></code></pre></div>
<p>Otherwise, it is possible to install the newest version from github
using:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># install the development version from GitHub</span>
<span class="co"># install.packages("devtools")</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"saezlab/CARNIVAL"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="inputs-and-outputs-of-carnival">Inputs and Outputs of CARNIVAL<a class="anchor" aria-label="anchor" href="#inputs-and-outputs-of-carnival"></a>
</h3>
<p>The input for CARNIVAL consists of:</p>
<ul>
<li><p>A prior knowledge network (PKN) comprises a list of signed and
directed interactions between signalling proteins. (Required)</p></li>
<li><p>Inferred transcription factor activities which can be inferred
from GEX data using <a href="https://github.com/saezlab/dorothea" class="external-link">DoRothEA</a>.
(Required)</p></li>
<li><p>A list of target of perturbations (drugs, diseases, etc.) with or
without their effects on signalling proteins. (Optional)</p></li>
<li><p>Inferred pathway scores representing signalling pathway
activities from GEX data using <a href="https://github.com/saezlab/progeny" class="external-link">PROGENy</a>
(Optional)</p></li>
</ul>
<p>The outcome of CARNIVAL includes the list of identified networks that
fitted to the provided experimental data as well as the predicted
activities of signalling proteins in the networks whether they are up-
or down-regulated.</p>
</div>
</div>
<div class="section level2">
<h2 id="running-carnival">Running CARNIVAL<a class="anchor" aria-label="anchor" href="#running-carnival"></a>
</h2>
<p>To obtain the list of tutorials/vignettes of the CARNIVAL package,
user can start with typing the following commmand on R-console:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/vignette.html" class="external-link">vignette</a></span><span class="op">(</span><span class="st">"CARNIVAL-vignette"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p><a href="https://pubs.rsc.org/en/content/articlehtml/2015/ib/c4ib00294f" class="external-link">Melas
et al.</a>:</p>
<blockquote>
<p>Melas IN, Sakellaropoulos T, Iorio F, Alexopoulos L, Loh WY,
Lauffenburger DA, Saez-Rodriguez J, Bai JPF. (2015). Identification of
drug-specific pathways based on gene expression data: application to
drug induced lung injury. <em>Integrative Biology</em>, Issue 7, Pages
904-920, <a href="https://doi.org/10.1039/C4IB00294F" class="external-link uri">https://doi.org/10.1039/C4IB00294F</a>.</p>
</blockquote>
<p><a href="https://github.com/saezlab/dorothea" class="external-link">DoRothEA -
Garcia-Alonso et al.</a>:</p>
<blockquote>
<p>Garcia-Alonso L, Holland CH, Ibrahim MM, Turei D, Saez-Rodriguez J.
(2018). Benchmark and integration of resources for the estimation of
human transcription factor activities. <em>Genome Research</em> 2019.
29: 1363-1375, <a href="https://doi.org/10.1101/gr.240663.118" class="external-link uri">https://doi.org/10.1101/gr.240663.118</a></p>
</blockquote>
<p><a href="https://github.com/saezlab/progeny" class="external-link">PROGENy - Schubert et
al.</a>:</p>
<blockquote>
<p>Schubert M, Klinger B, Klünemann M, Sieber A, Uhlitz F, Sauer S,
Garnett MJ, Blüthgen N, Saez-Rodriguez J. (2018). Perturbation-response
genes reveal signaling footprints in cancer gene expression. <em>Nature
Communication</em>, Issue 9, Nr. 20. <a href="https://doi.org/10.1038/s41467-017-02391-6" class="external-link uri">https://doi.org/10.1038/s41467-017-02391-6</a>.</p>
</blockquote>
</div>
<div class="section level2">
<h2 id="studies-where-carnival-was-used">Studies where CARNIVAL was used<a class="anchor" aria-label="anchor" href="#studies-where-carnival-was-used"></a>
</h2>
<blockquote>
<p>Buhl E.M. et al. <a href="https://www.embopress.org/doi/10.15252/emmm.201911021" class="external-link">Dysregulated
mesenchymal PDGFR‐β drives kidney fibrosis</a> <em>EMBO Mol Med
(2020)e11021 doi: 10.15252/emmm.201911021</em></p>
</blockquote>
<blockquote>
<p>Binenbaum I. et. al. <a href="https://link.springer.com/content/pdf/10.1007/s10522-020-09866-y.pdf" class="external-link">Bioinformatic
framework for analysis of transcription factor changes as the molecular
link between replicative cellular senescence signaling pathways and
carcinogenesis</a> <em>Biogerontology doi:
10.1007/s10522-020-09866-y</em></p>
</blockquote>
</div>
<div class="section level2">
<h2 id="acknowledgement">Acknowledgement<a class="anchor" aria-label="anchor" href="#acknowledgement"></a>
</h2>
<p>CARNIVAL has been developed as a computational tool to analyse -omics
data within the <a href="https://transqst.org" class="external-link">TransQST Consortium</a>
and <a href="https://www.h2020symbiosys.eu/" class="external-link">H2020 Symbiosys ITN
Training Network</a>.</p>
<p>“This project has received funding by the European Union’s H2020
program (675585 Marie-Curie ITN ‘‘SymBioSys’’) and the Innovative
Medicines Initiative 2 Joint Undertaking under grant agreement No
116030. The Joint Undertaking receives support from the European Union’s
Horizon 2020 research and innovation programme and EFPIA.”</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/CARNIVAL" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/saezlab/CARNIVAL/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/saezlab/CARNIVAL/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing CARNIVAL</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Enio Gjerga <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-3060-5786" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Panuwat Trairatphisan <br><small class="roles"> Author </small>  </li>
<li>Attila Gabor <br><small class="roles"> Maintainer </small>  </li>
<li>Olga Ivanova <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#stable" class="external-link"><img src="https://img.shields.io/badge/lifecycle-stable-brightgreen.svg" alt="Lifecycle: stable"></a></li>
<li><a href="https://bioconductor.org/checkResults/release/bioc-LATEST/CARNIVAL" class="external-link"><img src="http://www.bioconductor.org/shields/build/release/bioc/CARNIVAL.svg" alt="BioC status"></a></li>
<li><a href="https://bioconductor.org/checkResults/devel/bioc-LATEST/CARNIVAL" class="external-link"><img src="http://www.bioconductor.org/shields/build/devel/bioc/CARNIVAL.svg" alt="BioC devel status"></a></li>
<li><a href="https://codecov.io/gh/saezlab/CARNIVAL" class="external-link"><img src="https://codecov.io/gh/saezlab/CARNIVAL/branch/master/graph/badge.svg?token=5J039A2TP5" alt="codecov test coverage"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Enio Gjerga, Panuwat Trairatphisan, Attila Gabor, Olga
Ivanova.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
